fix RealtimePresence race; add StateEnum type

rjeczalik opened the pull request:
Channel may get attached after blocking Get is called on
RealtimePresence - when that happens, Get returns empty presence list.
This CL makes Get also wait for the channel until it's attached.

A StateEnum type was added to have connection/channel states human
readable when printed.

rjeczalik commented:
Hey @kouno! I've replaced `sync.WaitGroup` with just a `sync.Mutex` - it even simpler than `sync.Cond`. Please take a look :)

kouno commented on ably/realtime_presence.go:
```diff
@@ -11,25 +11,50 @@ import (
+	}
+}
+
+func (pres *RealtimePresence) stop() {
```
Why `stop` and not `close` like we do on the channel level?

rjeczalik replied:
Done.

kouno commented on ably/realtime_presence.go:
```diff
@@ -11,25 +11,50 @@ import (
+	return pres
+}
+
+func (pres *RealtimePresence) eventloop() {
```
`processIncomingAttachedEvent`? A bit verbose but `eventloop` is really not that descriptive.

rjeczalik replied:
Done.

kouno commented on ably/realtime_presence.go:
```diff
@@ -156,10 +199,15 @@ func (pres *RealtimePresence) Subscribe(states ...proto.PresenceState) (*Subscri
 	pres.subs.unsubscribe(true, sub, statesToKeys(states)...)
 }
 
+// Enter announces presence of the current client altogether with an enter
```
Not sure the `altogether` is needed.

rjeczalik replied:
Done.

kouno commented on ably/realtime_presence.go:
```diff
@@ -11,25 +11,50 @@ import (
+			pres.mtx.Lock()
+			if !pres.observedAttach && !pres.syncInProgress {
+				pres.syncMtx.Unlock() // channel is attached and there's no sync in progress
+				pres.observedAttach = true
```
I'm not sure what the role of `observedAttach` is.

If a channel is attached, then we get detached due to a network issue, this would stay true.

rjeczalik replied:
The `observedAttach` is there in order to differentiate two attach events - 1) an initial attach (right after the channel gets created), for which we want to block waiting for sync and 2) regular attach event.

But now I see that Get does not return error, when it's called on Detached / Failed channel. I'll address that and see if I can get rid of `observedAttach`.

rjeczalik replied:
Done.

kouno commented:
Looks ok. Can you address my few comments, especially the one about `observedAttach`?

kouno commented on ably/realtime_channel.go:
```diff
@@ -318,14 +318,14 @@ func (c *RealtimeChannel) Reason() error {
-		c.state.syncSet(StateChanAttached, nil)
-		c.queue.Flush()
 		if msg.Flags.Has(proto.FlagPresence) {
 			// since syncStart writes to RealtimePresence's internal WaitGroup
```
There is no WaitGroup anymore.

kouno commented on ably/realtime_presence.go:
```diff
@@ -25,11 +26,36 @@ type RealtimePresence struct {
+			}
+			pres.mtx.Unlock()
+		case StateChanDetached, StateChanFailed, StateChanClosed:
+			pres.syncMtx.Lock()
```
You are locking this in a completely separated thread but I don't see where it would get unlocked.

kouno commented on ably/realtime_presence.go:
```diff
@@ -25,11 +26,36 @@ type RealtimePresence struct {
+		switch event.State {
+		case StateChanAttached:
+			pres.mtx.Lock()
+			if !pres.syncInProgress {
```
TBH, it's difficult to get the whole picture from the code.

kouno commented on ably/realtime_presence.go:
```diff
@@ -25,11 +26,36 @@ type RealtimePresence struct {
+		event:   make(chan State, 1),
 		members: make(map[string]*proto.PresenceMessage),
 	}
+	pres.syncMtx.Lock() // initially we lock waiting for channel's attached state
```
I think you should avoid using short comments like this and explain it in a global comment at the top of your function.

kouno commented:
@rjeczalik I feel like this is not far off, but at the same time, every time I look at the locking system for incoming events, my head starts to spin :/. Especially since you added https://github.com/ably/ably-go/pull/35#discussion-diff-34672316R52.

rjeczalik commented:
@kouno This lock is direct consequence of the rewrite `sync.WaitGroup` -> `sync.Mutex` and I agree mutex operations are more confusing in the context. I take a step back look the the whole functionality and try to rewrite it to be more understandable.

rjeczalik commented:
Hey @kouno! Added `syncState` member, now I it's readable why extra blocking for the `initialState` state. I've also took opportunity and fixed erroneous rest test, which tried to marshal `*http.Request` - this is going to fail for go1.5.

Please take a look.

kouno commented:
:+1: a lot easier to follow thanks.

I added an issue for the channel state check, but I think we should close this PR which has been there for far too long. Can you sort out the incomplete message and merge?
